proposals/pr-contrib-guide.md

```diff
@@ -0,0 +1,315 @@
+
+Many pull requests come as unsolicited changes that often do not correspond to
+an open issue that has been discussed by the maintainers or the community. They
+often end up lingering with no communication from the maintainers to indicate
+the feasibility or appropriateness of the changes. As an example, [`aws-cli`
+pull request 2636](https://github.com/aws/aws-cli/pull/2636)  adds functionality
+to configure MFA. It adds or alters 231 lines of code across five files, and is
+a significant new feature. There is no open issue or discussion about the
+feature other than on the PR. It has no comments from AWS employees, other than
+to indicate that it is a “large” pull request, since 2017. Conversation
+continued for a year on the PR by interested users with no other interaction or
+guidance from AWS.
+
+A idea that has not been vetted by the community and the maintainers is
+difficult to approve; it's hard to have a conversation when it involves an idea
+proposal as well as code changes. It usually ends in an unmerged pull request or
+significant code revision, which is effort that was unnecessarily expended by
+the user and potentially the maintainer. We need a structured process for the
+life cycle of an idea that can culminate in a community contribution. We also
+need to prioritize issues against other requests from the community given the
+effort to contribute for users and review for maintainers.
+
+## Specification
+
+This section defines:
+
+* [Current process](#current-process)
+* [Proposed improvements](#proposed-improvements)
+
+### Current Process
+
+Customer feature requests are reviewed for general suitability and uniqueness,
+but are not actively commented on by maintainers to move the feature forward to
+an implementation. Pull requests are reviewed in an *ad hoc* fashion by various
+members of the maintainer team, but similarly are not actively processed to
+continue towards acceptance. This has contributed to the current status of 189
+pull requests open in the AWS CLI GitHub repository.
+
+### Proposed Improvements
+ 
+Maintainers that actively engage with the community contributors to identify
+suitable implementations for requested features and provide gudiance during the
+process will result in a higher percentage of successful community
+contributions. In addition, we will publish a contributing guide that describes
+the life cycle of a feature request that can culminate in a code contribution
+from the community through a GitHub pull request. The guide describes the triage
+and review of open feature requests with the goal of identifying issues where
+community contributions would be welcome. It defines what is required to move a
+feature request to be available for community contribution. It documents the
+review process and sets community expectations around communications to move a
+contribution forward. Figure A demonstrates a high level view of the proposed
+process.
```

Comment:
It still a little hard to follow what the exact process is mainly because the entire specification is one large section and not broken out. Specifically, it would be interesting if we made the following changes:

1. Introduce the notion of phases to the process. I'd imagine there would be three large phases (maybe we change the names of it):
   * Acceptance - This is the process in which an issue is used to track when an issue can be promoted to a state where would accept a PR
   * Implementation - This is process of implementing the contribution and making sure all of the prerequisites checks passed
   * Review - This is the cycle in which we work with the contributor to get the contribution merged.

2. Add a specification subsection for an overview of the process with the flow diagram and a description of these three phases. This would be the first part of the specification.

3. We have subsections that go into more detail about each of the phases

This hopefully should make it easier to break down how the entire process works.

Reply:
Mostly resolved - broken down into subsections for each 'phase'. Todo: updates to the figure, but resolving conversation for clarity.